lib/auth/grpcserver.go

```diff
@@ -1886,6 +1886,78 @@ func (g *GRPCServer) IsMFARequired(ctx context.Context, req *proto.IsMFARequired
 	return resp, nil
 }
 
+// GetSAMLConnector retrieves a SAML connector by name.
+func (g *GRPCServer) GetSAMLConnector(ctx context.Context, req *types.ResourceWithSecretsRequest) (*types.SAMLConnectorV2, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	samlConnector, err := auth.ServerWithRoles.GetSAMLConnector(ctx, req.Name, req.WithSecrets)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	samlConnectorV2, ok := samlConnector.(*types.SAMLConnectorV2)
+	if !ok {
+		return nil, trail.ToGRPC(trace.Errorf("encountered unexpected SAML connector type: %T", samlConnector))
+	}
+	return samlConnectorV2, nil
+}
+
+// GetSAMLConnectors retrieves all SAML connectors.
+func (g *GRPCServer) GetSAMLConnectors(ctx context.Context, req *types.ResourcesWithSecretsRequest) (*types.SAMLConnectorV2List, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	samlConnectors, err := auth.ServerWithRoles.GetSAMLConnectors(ctx, req.WithSecrets)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	samlConnectorsV2 := make([]*types.SAMLConnectorV2, len(samlConnectors))
+	for i, sc := range samlConnectors {
+		var ok bool
+		if samlConnectorsV2[i], ok = sc.(*types.SAMLConnectorV2); !ok {
+			return nil, trail.ToGRPC(trace.BadParameter("unexpected type %T", sc))
+		}
+	}
+	return &types.SAMLConnectorV2List{
+		SAMLConnectors: samlConnectorsV2,
+	}, nil
+}
+
+// UpsertSAMLConnector upserts a SAML connector.
+func (g *GRPCServer) UpsertSAMLConnector(ctx context.Context, samlConnector *types.SAMLConnectorV2) (*empty.Empty, error) {
+	auth, err := g.authenticate(ctx)
+	if err != nil {
+		return nil, trail.ToGRPC(err)
+	}
+	if err = services.ValidateSAMLConnector(samlConnector); err != nil {
+		return nil, trail.ToGRPC(err)
+	}
```

Comment:
Nit: This validation logic is not transport-specific.  Preferable to move it to the `ServerWithRoles` level.

Reply:
This is also being called by `IdentityService` already. It is a bit redundant, but it might be good to validate requests at the top of the call stack, what do you think? This is also being done on other requests, so I'll save this change for another PR. I hope to refactor validation across all resources/requests in 7.0 - https://github.com/gravitational/teleport/pull/5958

Reply:
I plan on refactoring this in a future PR, I'd prefer to keep things uniform for now (Users, Roles, and Access Requests all do validation here) - https://github.com/gravitational/teleport/pull/5958